packages/samples/headless-react/src/pages/SearchPage.tsx

```diff
@@ -356,6 +356,34 @@ export class SearchPage extends Component {
     this.unsubscribeExecuteTrigger();
   }
 
+  private get messageExpression() {
+    return buildQueryExpression({
+      operator: 'and',
+    })
+      .addStringField({
+        field: 'objecttype',
+        operator: 'isExactly',
+        values: ['Message'],
+      })
+      .toString();
+  }
+
+  private get confluenceExpression() {
+    return buildQueryExpression({operator: 'and'})
+      .addStringField({
+        field: 'connectortype',
+        operator: 'isExactly',
+        values: ['Confluence2Crawler'],
+      })
+      .addStringField({
+        field: 'documenttype',
+        operator: 'isExactly',
+        values: ['Space'],
+        negate: true,
+      })
+      .toString();
```

Comment:
So, this is what building strings with the builder pattern looks like.

I was discussing with @acote-coveo how to make it less verbose, and he suggested having multiple composable functions/classes. Using such a design, the above expression might look something like:

```
const {and, stringField, not} = buildQueryExpression();

and(
   stringField({field: 'connectortype', operator: 'isExactly', values: ['Confluence2Crawler']}),
   not(stringField({field: 'documenttype', operator: 'isExactly', values: ['Space']}))
)
```

We could also remove the objects
```
and(
   stringField('connectortype', 'isExactly', ['Confluence2Crawler']),
   not(stringField('documenttype', 'isExactly', ['Space']))
)
```

We can do the same with the builder pattern too.

One drawback of not using objects is if there are ever multiple optional properties (e.g. other than `negate`), then one would need to specify them all to specify the optional arg they care about. e.g.

```
stringField(field, operator, value, _something optional_, the optional thing I want to specify).
```


For this, @acote-coveo suggested using overloads to allow both positional args and a single object. Typescript allows this e.g.

```
export function buildStringField(field:string, operator: StringOperator, values: string[]): Part;
export function buildStringField(expression: StringFieldExpression): Part;

export function buildStringField(...args: [StringFieldExpression] | [string, StringOperator, string[]]): Part {
  const config: StringFieldExpression = args.length === 1 ? args[0] : {field: args[0], operator: args[1], values: args[2]}
  return {
      ...
  }
})
```

For docs, the first overload is what will appear in the parsed json.


So, to sum up, there doesn't seem to be any technical limitations with the composed approach.
Question is, which api would you enjoy using more? Builder function, or multiple composable functions?

Reply:
Of course, a part of me feels it's wrong to see such long query expressions, but I don't think it's actually a bad thing. If somebody's expression contains too many terms where it gets hard to read, they can extract some parts in other variables or functions easily.

I'm more worried about potential issues with using helper methods.

Reply:
Personally, I much prefer the builder pattern.

The main reason is that it's a very "classic" pattern from Java, and as such, if I was a user, I would instantly/naturally understand how to use it.

It's verbose and boring yeah, but most importantly it's * very * simple to understand. Boring is good.

It reads top to bottom, and I instantly know very well what I'll end up with at the end when calling `.toString()`

Versus this: 

```
and(
   stringField({field: 'connectortype', operator: 'isExactly', values: ['Confluence2Crawler']}),
   not(stringField({field: 'documenttype', operator: 'isExactly', values: ['Space']}))
)
```

I'm genuinely concerned that people would not actually understand how to use this system on their own without some documentation or copy pasting examples.

Reply:
Yea, api design is tough. Personally I found multiple small functions was easier to follow, but it's also less conventional. It seems like the team prefers the builder though, so we'll continue as is,